### PR TITLE
Added bollinger bands and moving average to candle/ohlc

### PR DESCRIPTION
**Pull request by andy-lee-eng**
Also reorganised the seriesColor stuff to make it a bit more reusable

image

### CODE REVIEW

**Comment by Ro4052** [comment]
Might have to change `colours` to `colors`.

File: packages/perspective-viewer-d3fc/src/js/charts/ohlcCandle.js [hunk 1]
```diff
@@ -16,22 +16,36 @@ import withGridLines from "../gridlines/gridlines";
...
 
 const isUp = d => d.closeValue >= d.openValue;
 
 function ohlcCandle(seriesCanvas) {
     return function(container, settings) {
-        const data = ohlcData(settings, filterDataByGroup(settings));
+        const srcData = ohlcData(settings, filterDataByGroup(settings));
 
-        const keys = data
+        const bollinger = fc.indicatorBollingerBands().value(d => d.openValue);
```
Comment (Ro4052) [inline packages/perspective-viewer-d3fc/src/js/charts/ohlcCandle.js:27]: Are there a lot of `bollinger` specific calculations we do even if we draw the ohlc or candle stick charts? I guess this only happens at the beginning.

File: packages/perspective-viewer-d3fc/src/js/series/seriesColors.js [hunk 2]
```diff
@@ -26,40 +26,69 @@ export function seriesColorsFromGroups(settings) {
...
+    let mapFunction = withOpacity;
 
-export function withoutOpacity(color) {
-    const lastComma = color.lastIndexOf(",");
-    const valueIndex = color.indexOf("(");
-    return valueIndex !== -1 && lastComma !== -1 ? `rgb(${color.substring(valueIndex + 1, lastComma)})` : color;
-}
+    const colors = () => {
+        if (defaultColours || domain.length > 1) {
+            const range = domain.length > 1 ? d3.schemeCategory10.map(mapFunction) : defaultColours;
```
Comment (Ro4052) [inline packages/perspective-viewer-d3fc/src/js/series/seriesColors.js]: Is `defaultColours` only supposed to be used if we have a domain of 1 or less?

**Review by mhooper-scottlogic — COMMENTED** [review]
It looks really good. My only comment would be that it might be worth moving the series creation part (series, bollingerAverageSeries, bollingerAreaSeries and mergedSeries) into a separate file under the series directory like we have for other charts. I didn't do this originally because the existing series was so simple but now it is more complex it might be nice to move it out to its own file.

The legend uses the upColor series, but the other two colorSeries are only referenced by the series creation statements so could be created inside the series file.

File: packages/perspective-viewer-d3fc/src/js/series/seriesColors.js (hunk 2 above)
Comment (andy-lee-eng) [reply packages/perspective-viewer-d3fc/src/js/series/seriesColors.js]: Yes, and it's only needed for canvas charts, and just avoids having a fallback in the chart code to style the colour if the colourSeries is `null`. For Svg charts, we fall back on styling with css instead.

Comment (Ro4052) [reply packages/perspective-viewer-d3fc/src/js/series/seriesColors.js]: 👍

**Review by mhooper-scottlogic — APPROVED** [review]

**Review by Ro4052 — APPROVED** [review]

**Pull request merged**